src/uetai/callbacks/image_monitor.py

```diff
@@ -1,136 +1,162 @@
 """image classifier callbacks"""
-from typing import Any, Dict, List, Optional
+import warnings
+from typing import Any, Dict, List, Optional, Union
 
-import torch
 from torch import Tensor
 from torchvision import transforms
-from pytorch_lightning import Trainer, Callback, LightningModule
-from pytorch_lightning.loggers.base import LightningLoggerBase
-from pytorch_lightning.utilities.warnings import rank_zero_warn
+
+import pytorch_lightning as pl
+from pytorch_lightning import Callback
 
 from uetai.logger import SummaryWriter
+from uetai.callbacks.utils import check_logger, trainer_finish_run
 
 try:
     import wandb
 except (ImportError, AssertionError):
     wandb = None
+    warnings.warn('Missing package `wandb`. Run `pip install wandb` to install it')
 
 
 class ImageMonitorBase(Callback):
-    """Base class for monitoring image data in a LightningModule.
+    """Base class for monitoring image data.
     """
     def __init__(
-        self, log_every_n_steps: int = None, label_mapping: Dict[int, str] = None
+        self,
+        on_step: bool = True,
+        on_epoch: bool = False,
+        log_every_n_steps: int = None,
+        log_n_element_per_epoch: int = None,
+        label_mapping: Dict[int, Any] = None,  # idx, name
     ):
         super().__init__()
-        self._log_every_n_steps: Optional[int] = log_every_n_steps
         self._log = False
-        self._trainer = Trainer
-        self._train_batch_idx: int
-        self._mapping = label_mapping
-
-    def on_train_start(
-        self, trainer: Trainer, pl_module: LightningModule
-    ) -> None:
-        self._log = self._check_logger(trainer.logger)
-        self._log_every_n_steps = self._log_every_n_steps or trainer.log_every_n_steps
+        self._trainer = pl.Trainer
+        self._on_step = on_step
+        self._on_epoch = on_epoch
+        self._label_mapping = label_mapping
+        self._log_every_n_steps = log_every_n_steps  # default is 50 in trainer.log_every_n_steps
+        self._log_n_element_per_epoch = log_n_element_per_epoch  # default log all
+        if self._on_epoch:
+            self._epoch: Dict[str, List] = {'images': [], 'ground_truths': [], 'predictions': []}
+            self._on_step = False
+
+    def on_train_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
         self._trainer = trainer
+        self._log = check_logger(trainer.logger)
+        self._log_every_n_steps = self._log_every_n_steps or trainer.log_every_n_steps
+        if not (self._on_step or self._on_epoch):
+            self._log = False
 
     def on_train_batch_end(
-        self, trainer: Trainer,
-        pl_module: LightningModule,
+        self, trainer: "pl.Trainer",
+        pl_module: "pl.LightningModule",
```

Comment:
Why do we need lightning module here?

Reply:
I'm overriding the `pytorch_lightning.Callbacks` method in this line.